lib/Listeners/Files/PreparingMemberSendMail.php

```diff
@@ -0,0 +1,148 @@
+use OCA\Circles\Events\PreparingCircleMemberEvent;
+use OCA\Circles\Exceptions\FederatedItemException;
+use OCA\Circles\Exceptions\RemoteInstanceException;
+use OCA\Circles\Exceptions\RemoteNotFoundException;
+use OCA\Circles\Exceptions\RemoteResourceNotFoundException;
+use OCA\Circles\Exceptions\RequestBuilderException;
+use OCA\Circles\Exceptions\UnknownRemoteException;
+use OCA\Circles\Model\Member;
+use OCA\Circles\Service\ConfigService;
+use OCA\Circles\Service\ContactService;
+use OCA\Circles\Service\ShareTokenService;
+use OCA\Circles\Service\ShareWrapperService;
+use OCP\EventDispatcher\Event;
+use OCP\EventDispatcher\IEventListener;
+use OCP\Security\IHasher;
+
+/**
+ * Class PreparingMemberSendMail
+ *
+ * @package OCA\Circles\Listeners\Files
+ */
+class PreparingMemberSendMail implements IEventListener {
+	use TStringTools;
+	use TNC22Logger;
+
+
+	/** @var IHasher */
+	private $hasher;
+
+	/** @var ShareWrapperService */
+	private $shareWrapperService;
+
+	/** @var ShareTokenService */
+	private $shareTokenService;
+
+	/** @var ConfigService */
+	private $configService;
+
+	/** @var ContactService */
+	private $contactService;
+
+
+	/**
+	 * AddingMember constructor.
+	 *
+	 * @param IHasher $hasher
+	 * @param ShareWrapperService $shareWrapperService
+	 * @param ShareTokenService $shareTokenService
+	 * @param ContactService $contactService
+	 * @param ConfigService $configService
+	 */
+	public function __construct(
+		IHasher $hasher,
+		ShareWrapperService $shareWrapperService,
+		ShareTokenService $shareTokenService,
+		ContactService $contactService,
+		ConfigService $configService
+	) {
+		$this->hasher = $hasher;
+		$this->shareWrapperService = $shareWrapperService;
+		$this->shareTokenService = $shareTokenService;
+		$this->contactService = $contactService;
+		$this->configService = $configService;
+
+		$this->setup('app', Application::APP_ID);
+	}
+
+
+	/**
+	 * @param Event $event
+	 *
+	 * @throws RequestBuilderException
+	 * @throws FederatedItemException
+	 * @throws RemoteInstanceException
+	 * @throws RemoteNotFoundException
+	 * @throws RemoteResourceNotFoundException
+	 * @throws UnknownRemoteException
+	 */
+	public function handle(Event $event): void {
+		if (!$event instanceof PreparingCircleMemberEvent
+			|| !$this->configService->enforcePasswordOnSharedFile()) {
+			return;
+		}
+
+		$member = $event->getMember();
+		if ($member->getUserType() === Member::TYPE_CIRCLE) {
+			$members = $member->getBasedOn()->getInheritedMembers();
+		} else {
+			$members = [$member];
+		}
+
+		$federatedEvent = $event->getFederatedEvent();
+		$clearPasswords = $federatedEvent->getInternal()->gArray('clearPasswords');
+		$hashedPasswords = $federatedEvent->getParams()->gArray('hashedPasswords');
+
+		foreach ($members as $member) {
+			if (($member->getUserType() !== Member::TYPE_MAIL
+				 && $member->getUserType() !== Member::TYPE_CONTACT)
+				|| array_key_exists($member->getSingleId(), $clearPasswords)
```

Comment:
Maybe add a comment here on what the conditions are for? From my understanding, it's "Only process members of TYPE_MAIL or TYPE_CONTACT, but ignore those who already have a password set", right?

Reply:
comments added